---
# _Introduction to Pandas_
---
## _Quiz 2 (20 Sep)_

In [1]:
import pandas as pd
sma_data = pd.read_csv('Standard_Metropolitan_Areas_Data-data.csv')

### Q1: 
sma_data is a pandas dataframe. What is the shape of the dataframe 'sma_data'?

In [2]:
sma_data.shape

(99, 10)

### Q2: 
How many entries are there in the dataframe sma_data?

In [3]:
len(sma_data)

99

### Q3: 
Which of the following method will help you display the last five observations from the dataframe?
> a) head()  
> **b) tail()**

In [4]:
sma_data.head()

land_area  percent_city  percent_senior  physicians  hospital_beds  \
0       1384          78.1            12.3       25627          69678   
1       3719          43.9             9.4       13326          43292   
2       3553          37.4            10.7        9724          33731   
3       3916          29.9             8.8        6402          24167   
4       2480          31.5            10.5        8502          16751   

   graduates  work_force  income  region  crime_rate  
0       50.1      4083.9   72100       1       75.55  
1       53.9      3305.9   54542       2       56.03  
2       50.6      2066.3   33216       1       41.32  
3       52.2      1966.7   32906       2       67.38  
4       66.1      1514.5   26573       4       80.19

In [5]:
sma_data.tail()

land_area  percent_city  percent_senior  physicians  hospital_beds  \
94       1511          38.7            10.7         348           1093   
95       1543          39.6             8.1         159            481   
96       1011          37.8            10.5         264            964   
97        813          13.4            10.9         371           4355   
98        654          28.8             3.9         140           1296   

    graduates  work_force  income  region  crime_rate  
94       50.4       127.2    1452       4       70.66  
95       30.3        80.6     769       3       36.36  
96       70.7        93.2    1337       3       60.16  
97       58.0        97.0    1589       1       36.33  
98       55.1        66.9    1148       3       68.76

### Q4:
Select the correct statement about the data types of different columns.

> a) The data type of land_area is integer  
> b) There are 5 columns whose data types is float   
> c) There are 5 columns whose data types is integer   
> **d) All of the above are correct**

In [6]:
sma_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99 entries, 0 to 98
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   land_area       99 non-null     int64  
 1   percent_city    99 non-null     float64
 2   percent_senior  99 non-null     float64
 3   physicians      99 non-null     int64  
 4   hospital_beds   99 non-null     int64  
 5   graduates       99 non-null     float64
 6   work_force      99 non-null     float64
 7   income          99 non-null     int64  
 8   region          99 non-null     int64  
 9   crime_rate      99 non-null     float64
dtypes: float64(5), int64(5)
memory usage: 7.9 KB


---
## _Quiz 3 (23 Sep)_

### Q1: 
Use the index-based selection technique to select the 10th observation from data frame sma_data. What is the crime_rate for this observation? 
> a) 30.51  
> **b) 55.30**  
> c) 54.16  
> d) 52.73

In [7]:
sma_data.iloc[9]

land_area          4647.0
percent_city         31.5
percent_senior        9.2
physicians         3916.0
hospital_beds     12815.0
graduates            65.1
work_force         1032.2
income            14542.0
region                2.0
crime_rate           55.3
Name: 9, dtype: float64

### Q2. 
Select the physicians column. Get the last value from physicians column.
> **a) 140**  
> b) 135  
> c) 145  
> d) 142  

In [8]:
sma_data.loc[:, 'physicians']

0     25627
1     13326
2      9724
3      6402
4      8502
      ...  
94      348
95      159
96      264
97      371
98      140
Name: physicians, Length: 99, dtype: int64

### Q3. 
Select the records with index labels (numerical positions) - 1, 3, 5, 7, 9 and 13 and columns - 'land_area', 'work_force', 'income', 'region' and 'crime_rate'. Assign the result to a variable sample_data1. Select the correct statement about this sample_data1. [Hint: Use label-based selection technique]
> a) The crime rates are less than 70 for all the observations in the dataframe sample_data1  
> b) The observation indexed at 13 has income of 12591  
> c) **Both of the above are correct**  
> d) None of the above are correct

In [9]:
sample_data1 = sma_data.loc[[1, 3, 5, 7, 9, 13], ['land_area', 'work_force', 'income', 'region', 'crime_rate']]
sample_data1

land_area  work_force  income  region  crime_rate
1        3719      3305.9   54542       2       56.03
3        3916      1966.7   32906       2       67.38
5        2815      1541.9   25663       3       58.48
7        6794      1272.7   18221       3       64.88
9        4647      1032.2   14542       2       55.30
13        782       915.2   12591       4       63.20

### Q4: 
Select the observations for which the region is equal to 2. Assign the result to a variable sample_data2. How many observations are there in the dataframe sample_data2? [Hint: sma_data.region is equal to?]
> **a) 25**  
> b) 10  
> c) 43  
> d) 63

In [10]:
sample_data2 = sma_data[sma_data.region == 2]
len(sample_data2)

25

---
## _Quiz 4 (23 Sep)_

### Q1: 
Use mean() method on the dataframe sma_data and select the correct statements.
> a) The mean of the column land_area is in the range 2615 to 2616  
> b) The mean of the column income is in the range 6762 to 6763  
> **c) Both are correct**  
> d) None of the above are correct  

In [11]:
sma_data.mean()

land_area         2615.727273
percent_city        42.518182
percent_senior       9.781818
physicians        1828.333333
hospital_beds     6345.868687
graduates           54.463636
work_force         449.366667
income            6762.505051
region               2.494949
crime_rate          55.643030
dtype: float64

### Q2:
Use unique() for the column region. How many unique values are there in this column?
> a) 1  
> b) 2  
> c) 3  
> **d) 4**

In [12]:
sma_data.region.unique()

array([1, 2, 4, 3], dtype=int64)

### Q3:
Select the correct statement about the sma_data. [Hint: Summary function - describe()]
> a) The minimum income is 769.0  
> b) The maximum crime rate is 85.62  
> c) The median land area is 1951.0  
> d) There are 99 non-null values in work_force column  
> **e) All of the above are correct**

In [13]:
sma_data.describe()

land_area  percent_city  percent_senior    physicians  \
count     99.000000     99.000000       99.000000     99.000000   
mean    2615.727273     42.518182        9.781818   1828.333333   
std     3045.826210     17.348277        2.524547   3192.199763   
min       47.000000     13.400000        3.900000    140.000000   
25%     1408.000000     30.100000        8.350000    459.000000   
50%     1951.000000     39.500000        9.700000    774.000000   
75%     2890.500000     52.600000       10.750000   1911.500000   
max    27293.000000    100.000000       21.800000  25627.000000   

       hospital_beds  graduates   work_force        income     region  \
count      99.000000  99.000000    99.000000     99.000000  99.000000   
mean     6345.868687  54.463636   449.366667   6762.505051   2.494949   
std      9136.202716   7.773286   610.990885  10393.349660   1.013921   
min       481.000000  30.300000    66.900000    769.000000   1.000000   
25%      2390.000000  50.250000   150.300000   2003.000000   2.000000   
50%      3472.000000  54.000000   257.200000   3510.000000   3.000000   
75%      6386.500000  58.300000   436.500000   6283.500000   3.000000   
max     69678.000000  72.800000  4083.900000  72100.000000   4.000000   

       crime_rate  
count   99.000000  
mean    55.643030  
std     13.470943  
min     23.320000  
25%     46.115000  
50%     56.060000  
75%     63.860000  
max     85.620000

### Q4: 
Select the observations from the dataframe sma_data where the region is equal to 3. Name this selected data as sample_data1. select the correct options about sample_data1.
> 1) There are 36 observations  
> 2) The mean crime rate is in the range 58.0 to 58.4  
> 3) There are 30 non - null values in the column graduates  
> 4) All of the above are correct  
> **5) Only 1 and 2 are correct**

In [14]:
sample_data1 = sma_data[sma_data.region == 3]
sample_data1.describe()

land_area  percent_city  percent_senior   physicians  hospital_beds  \
count    36.000000     36.000000       36.000000    36.000000      36.000000   
mean   2485.972222     46.172222        9.283333  1313.722222    4962.777778   
std    1653.813800     19.101019        3.458199  1459.654903    3879.692625   
min     654.000000     14.500000        3.900000   140.000000     481.000000   
25%    1517.000000     31.350000        7.600000   509.750000    2483.500000   
50%    2018.000000     47.000000        8.850000   767.000000    3876.000000   
75%    2911.000000     59.725000        9.975000  1583.500000    5811.750000   
max    8360.000000     90.700000       21.800000  7340.000000   16941.000000   

       graduates   work_force        income  region  crime_rate  
count  36.000000    36.000000     36.000000    36.0   36.000000  
mean   51.002778   343.797222   4806.861111     3.0   58.265556  
std     7.328164   354.312228   5452.647656     0.0   10.155822  
min    30.300000    66.900000    769.000000     3.0   36.360000  
25%    46.950000   146.200000   1940.750000     3.0   52.447500  
50%    50.950000   230.550000   2976.500000     3.0   58.725000  
75%    55.175000   366.950000   4829.500000     3.0   64.922500  
max    70.700000  1541.900000  25663.000000     3.0   76.350000

### Q5:
Count the occurence of unique values in the column region of the dataframe sma_data. Select the correct statements. [Hint: value_counts()]
> 1) The maximum occurring region is 3  
> 2) The minimum occurring region is 2  
> 3) The region 4 is occurring 17 times  
> **4) Both 1 and 3 are correct**   
> 5) Both 1 and 2 are correct

In [15]:
sma_data.region.value_counts()

3    36
2    25
1    21
4    17
Name: region, dtype: int64

---
## _Quiz 5 (26 Sep)_

### Q1: 
Sort the dataframe by crime rate in descending order. [Note: The default sorting is in ascending order].    
Assign the sorted data to a variable sorted_data1. The first five observations of the dataframe sorted_data1 has the same region.   
Which region is that?
> **a) 4**   
> b) 3  
> c) 2  
> d) 1  
> e) 0

In [16]:
sorted_data1 = sma_data.sort_values('crime_rate', ascending=False)[:5]
sorted_data1['region'][:5]

20    4
74    4
53    4
4     4
67    4
Name: region, dtype: int64

### Q2:
Which method is used to rename a column/index?
> **a) rename()**  
> b) Rename()  
> c) ReName()  
> d) Both 1 and 2

### Q3:
Which of the following set of code will help you to rename the column physicians in the original dataframe?
> **a) sma_data.rename(columns = {'physicians': 'num_physicians'}, inplace = True)**  
> b) sma_data.rename(columns = {'physicians': 'num_physicians'}, inplace = False)  
> c) sma_data.rename(columns = {'physicians': 'num_physicians'})  
> d) Both 1 and 3

In [17]:
sma_data = pd.read_csv('Standard_Metropolitan_Areas_Data-data.csv')
sma_data.rename(columns = {'physicians': 'num_physicians'}, inplace = True)
sma_data.head(1)

land_area  percent_city  percent_senior  num_physicians  hospital_beds  \
0       1384          78.1            12.3           25627          69678   

   graduates  work_force  income  region  crime_rate  
0       50.1      4083.9   72100       1       75.55

In [18]:
sma_data = pd.read_csv('Standard_Metropolitan_Areas_Data-data.csv')
sma_data.rename(columns = {'physicians': 'num_physicians'}, inplace = False)
sma_data.head(1)

land_area  percent_city  percent_senior  physicians  hospital_beds  \
0       1384          78.1            12.3       25627          69678   

   graduates  work_force  income  region  crime_rate  
0       50.1      4083.9   72100       1       75.55

In [19]:
sma_data = pd.read_csv('Standard_Metropolitan_Areas_Data-data.csv')
sma_data.rename(columns = {'physicians': 'num_physicians'})
sma_data.head(1)

land_area  percent_city  percent_senior  physicians  hospital_beds  \
0       1384          78.1            12.3       25627          69678   

   graduates  work_force  income  region  crime_rate  
0       50.1      4083.9   72100       1       75.55

---